updating documentation for including semantic information in CDL

anandkp92 opened the pull request.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,75 @@ Therefore, tools that process CDL can infer the following information:
+ * Inputs
+ * Outputs
+ * Parameters
+ * CDL blocks: both elementary blocks (:numref:`sec_ele_blo`), and composite blocks (:numref:`sec_com_blo`)
```
Elementary blocks are in an immutable library, which should be write protected. Hence we cannot add an annotation. We can however instantiate this in a composite block and add there the annotation. That should suffice in my view. Therefore, I think the "elementary block" needs to be removed from this list.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,75 @@ Therefore, tools that process CDL can infer the following information:
 
-* elementary blocks (:numref:`sec_ele_blo`), and
-* composite blocks (:numref:`sec_com_blo`).
+* The top-level CDL block declaration shall house the `metadataLanguageDefinition` within the ``__semantic`` annotation. ``metadataLanguageDefinition`` will be used to define the different types of tags that are used throughout the CDL block and relevant information about the same. It will follow the following syntax:
```
You need to add two quotes in metadataLanguageDefinition, see https://github.com/lbl-srg/obc/blob/issue133_semanticDocumentation/specification/source/cdl.rst#semantic-information for rendering. There are also other occurrences of this.

anandkp92 replied:
I corrected it here, but I couldn't find other instances - so let me know if you encounter errors while rendering.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,75 @@ Therefore, tools that process CDL can infer the following information:
-adding a tag to a ``parameter`` declaration.]
+.. code-block:: modelica
 
+   annotation (__semantic(metadataLanguageDefinition="<metadataLanguage> [version] [format]" "<description or URL>"))
```
I am not opposed to adding the optional ``[format]`` but isn't the format always implied by `metadataLanguage`?

anandkp92 replied:
Not always - For example, Haystack alone has multiple formats: JSON, JSON+LD, Zinc, Turtle etc. (https://project-haystack.org/example)

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,75 @@ Therefore, tools that process CDL can infer the following information:
-to an ahu3. CDL, however, has a different syntax to declare such dependencies:
-In CDL, through the ``connect(whitehouse.ahu3.TSup, ...)`` statement,
-a tool can infer what upstream component sends the input signal.]
+* Tools that process CDL can interpret the ``__semantic`` annotation, but for control purposes CDL will ignore it. [This avoids potential conflict for entities that are declared differently across these metadata standards and CDL, and may be conflicting.]
```
Above that line is an (old) paragraph
```
The metadata defined in the top-level bl annotations will be accessible to elements (inputs, outputs etc.) defined within the sequence. Additionally, semantic information attached to an element can refer to elements that were defined earlier in the sequence. This way, the semantic information attached to an element can refer to another element within the sequence that was defined previously.
```
Please rewrite this and use instance rather than element when you mean an instance of an input, output, block etc. as we do not have a term called "element" in the CDL description. (We should always use the precise technical terms to avoid confusing the new reader.)

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,75 @@ Therefore, tools that process CDL can infer the following information:
+`Brick <https://brickschema.org/>`_, `Project Haystack <https://project-haystack.org/>`_ and `ASHRAE 223p <https://www.ashrae.org/about/news/2018/ashrae-s-bacnet-committee-project-haystack-and-brick-schema-collaborating-to-provide-unified-data-semantic-modeling-solution>`_. CDL allows, but does not require, use of the Brick or Haystack or any other tagging scheme.
 
-CDL allows to add tags to declarations that instantiate
+* Semantic information will be included within the CDL ``annotation`` keyword, using the ``__semantic`` annotation. Every element in CDL can have an annotation and correspondingly, semantic information
```
I know we agree to use `__semantic` but this can cause a name conflict if some other library also adds `__semantic` but using their own definition. The specification of vendor annotation is here:
https://specification.modelica.org/v3.4/Ch18.html
Based on that, on second thoughts I think we should call it something like `__Buildings(semantic(...))`, where `Buildings` is the name of the library, which I chose to avoid putting `__LBNL(...)` into the standard. This should keep us clear of name clashes, and `modelica-json` can search for `__"name of library"(semantic(...))`. I also selected `__Buildings` as I think use of `__ASHRAE_...` would need permission, and is not suited for other semantic annotation such as for equipment that may be in Brick or Haystack, or plain text.

So the formal specification would be
```
__"name of the library"(semantic(...))
```
where "name of the library" is the name of the top-level package, such as `Buildings` for sequences that are stored in the `Buildings` library.  If a user develops an own library of sequences that may be called "MyControlLibrary" then the annotation can either be ``__MyControlLibrary(semantic(...))`` or ``__Buildings(semantic(...))`` because a tool, such as modelica-json, that parses the semantic information can look for both of these strings. Hence, sequences copied from a library such as `Buildings` into a user library such as `MyControlLibrary` can retain the semantic information as-is, or they can optionally rename it and add their own semantic information.

mwetter replied:
@anandkp92 : Actually, `__MyControlLibrary(semantic(...))` can give the same name clash as we want to avoid with abandoning `__semantic`. So I think the options should be either of these two
- `__Buildings(semantic(...))` in the Buildings library, and also allow
- `__CDL(semantic(...))`.

We can then use `__CDL` in the CDL sequences, and `__Buildings` everywhere else, and modelica-json will treat them the same. I think that will work unless you have concerns, because we are in control of both names, and the ASHRAE standard would not need to be tied to the Buildings library.

anandkp92 replied:
I have updated the specification to use __Buildings and __cdl (to be consistent with rest of CDL annotations)

mwetter reviewed (changes requested):
@anandkp92 : This looks good but I have a few changes, including a name change for `__semantic`. Please see the inline comments.

anandkp92 commented:
@mwetter I've made the requested changes. Please let me know if I missed anything.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
-Brick (https://brickschema.org/) and Haystack (https://project-haystack.org/).
-CDL allows, but does not require, use of the Brick or Haystack tagging scheme.
+The buildings industry uses different metadata schemes such as
+`Brick <https://brickschema.org/>`_, `Project Haystack <https://project-haystack.org/>`_ and `ASHRAE 223p <https://www.ashrae.org/about/news/2018/ashrae-s-bacnet-committee-project-haystack-and-brick-schema-collaborating-to-provide-unified-data-semantic-modeling-solution>`_. CDL allows, but does not require, use of the Brick or Haystack or any other tagging scheme.
```
Rewrite that 223P may be used in the future. I don't think we can write that "the buildings industry uses ... 223P" if it is still in development.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
 .. _sec_tag_pro:
 
-Tagged Properties
+Semantic Information
```
Correct translation errors:
```
/home/mwetter/proj/doe/obc/lbl-srg/obc/specification/source/cdl.rst:1692: WARNING: Title underline too short.

Semantic Information
.................
/home/mwetter/proj/doe/obc/lbl-srg/obc/specification/source/cdl.rst:1692: WARNING: Title underline too short.

Semantic Information
.................
../lbl-srg/obc/specification/source/cdl.rst:1698: WARNING: Bullet list ends without a blank line; unexpected unindent.

```

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
+ * Composite blocks (:numref:`sec_com_blo`)
+ * Packages
+ 
+* All semantic information will be included under the ``semantic`` section within the ``__Building`` and/or ``__cdl`` annotations. 
```
I don't understand why line 1704 and below are also bullets. Is that intentional and if so, what is the paragraph structure?

anandkp92 replied:
I kept bullets+paragraphs because it looked more organized, but I've changed it to only paragraphs.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
+   annotation (__cdl(semantic(<semantic information>)))
+   annotation (__Buildings(semantic(<semantic information>)))
+   
+* The top-level block declaration shall house the ``metadataLanguageDefinition`` within the  ``semantic`` section. ``metadataLanguageDefinition`` will be used to define the different types of tags that are used throughout the CDL block and relevant information about the same. It will follow the following syntax:
```
Better explain what is meant by "relevant information about the same"?

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
-where ``__cdl__annotation`` is the annotation for CDL.
-
-For Brick, the ``__cdl_annotation`` is
+* ``metadataLanguageDefinition`` could be semantic data standards such as  Brick, Project-Haystack, ASHRAE 223p etc. or it could just be a natural language description. The optional ``version`` represents the version of the metadataLangauge used in a particular sequence and the optional ``format``  represents the format this will be expressed in.  The format will be expressed using  `MIME types < ...
```
Remove "standards" as none of them are a standard (yet).

mwetter replied:
Also, use quotes for ``metadataLangauge``

mwetter replied:
When stating an example, put it within brackes such as
```
[
...
]
```
See item 2 in http://obc.lbl.gov/specification/conventions.html

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
+  annotation (__Buildings(semantic(metadataLanguage="<metadataLanguage> [version] [format]" "<metadata>")));
 
-For example, the Brick specification
+* Annotations attached to the top-level block can also be used to define block level information about the semantic data. A top-level block is one that encompasses a particular sequence or a model. For example, in Brick 1.3, it can be used to define prefixes. Example:
```
What does "attached to the top-level mean"? Better write something like "Semantic information declared in the block-level annotation ...".
The use of "top-level" is problematic as a block may be at a top-level, but if used within another composite block, it is no longer at the top-level. 
Also, a block cannot encompass (contain?) a model as this is not valid Modelica.

mwetter replied:
I don't understand the introduction to the example. You wrote "in Brick 1.3" but then the example is for Brick 1.3 and for Project-Haystack. Either simplify the example to only use Brick, or adjust the introduction.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -1688,118 +1688,97 @@ Therefore, tools that process CDL can infer the following information:
+   metadataLanguage="Project-Haystack 3.9.12 application/ld+json" "url=https://url/to/model.json")));
 
-For Haystack, the ``__cdl_annotation`` is
+* Example of CDL semantic information is shown below. The metadataLanguage "Brick 1.3 text/turtle" has been defined earlier in a top-level block and hence can be used to include semantic information. Similarly, "Project-Haystack 3.9.12 application/ld+json" and "Text" have been defined earlier and that makes for a valid syntax
```
Does it really need to be top-level or will the `metadataLanguage` be taken from the first instance of a block that is at a higher hierarchy than the current instance? For example, if A extends B, and B extends C, and A and B contain `metadataLanguage`, does A override B, or B override A, or is it an error if their `metadataLanguage` are not the same?

anandkp92 replied:
@mwetter: This is a good point that you bring up that we must probably be explicit in the standard. I need to think a bit on this and get back to you.

mwetter reviewed (changes requested):
Please see inline comments.

anandkp92 commented:
@mwetter I've addressed your comments and added a note about inherited instances and higher level instance declaration. Please review and let me know if you have comments.

mwetter reviewed (changes requested):
@anandkp92 : Let's discuss on our next coordination meeting. This should get much more prescriptive and clearer. Please see my notes in bold font.

anandkp92 commented:
@mwetter I've made the necessary changes and added the information we discussed. It's ready for your review now.

mwetter commented:
@anandkp92 : please let me know when you are finished revising the section. And note that there are formatting issues reported by sphinx.

anandkp92 commented:
@mwetter ready for your review. Thanks!
